File: scm-java/src/main/scheme/primitives/PrimitiveStringMatches.java

package scheme.primitives;

import scheme.*;

import java.util.regex.*;
import java.util.concurrent.ConcurrentHashMap;

public class PrimitiveStringMatches extends Primitive {
    private static final ConcurrentHashMap<String, Pattern> cache = new ConcurrentHashMap<>();
    private static Pattern get(String pattern) {
        Pattern p = cache.get(pattern);
        if (p != null) return p;
        p = Pattern.compile(pattern);
        if (cache.size() > 256) cache.clear();
        cache.putIfAbsent(pattern, p);
        return p;
    }

    @Override
    public String name() {
        return "string-matches";
    }

    @Override
    public String info() {
        return "Syntax: (string-matches s pattern)\n" +
               "Library: (scm string)\n" +
               "Description: Matches the string s against the regular expression pattern. Returns a list of match strings (the full match followed by any groups) if successful, or #f if there is no match.\n" +
               "Example:\n" +
               "  (string-matches \"hello\" \"hel+o\") => (\"hello\")\n" +
               "  (string-matches \"abc123\" \"([a-z]+)([0-9]+)\") => (\"abc123\" \"abc\" \"123\")\n" +
               "  (string-matches \"hello\" \"xyz\") => #f";
    }
    
    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 2, 2);
        var s = new String(Value.asString(arguments[0]));
        var pattern = get(new String(Value.asString(arguments[1])));
        var matcher = pattern.matcher(s);
        if (matcher.find()) {
            var groups = new Object[matcher.groupCount() + 1];
            for (var i = 0; i <= matcher.groupCount(); i++) {
                groups[i] = matcher.group(i).toCharArray();
            }
            return Pair.list(groups);
        }
        return Value.F;
    }
}

File: scm-csharp/scheme/primitives/PrimitiveStringMatches.cs
using System.Collections.Concurrent;
using System.Text.RegularExpressions;

namespace scheme;

public class PrimitiveStringMatches : Primitive
{
    private static readonly ConcurrentDictionary<string, Regex> cache = new();
    private static Regex Get(string pattern)
    {
        if (cache.TryGetValue(pattern, out var r)) return r;
        var fresh = new Regex(pattern);
        if (cache.Count > 256) cache.Clear();
        cache.TryAdd(pattern, fresh);
        return fresh;
    }

    public override string Name()
    {
        return "string-matches";
    }

    public override string Info()
    {
        return
            "Syntax: (string-matches s pattern)\n" +
            "Library: (scm string)\n" +
            "Description: Matches the string s against the regular expression pattern. Returns a list of match strings (the full match followed by any groups) if successful, or #f if there is no match.\n" +
            "Example:\n" +
            "  (string-matches \"hello\" \"hel+o\") => (\"hello\")\n" +
            "  (string-matches \"abc123\" \"([a-z]+)([0-9]+)\") => (\"abc123\" \"abc\" \"123\")\n" +
            "  (string-matches \"hello\" \"xyz\") => #f";
    }
    
    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 2, 2);
        var s = new String(Value.AsString(arguments[0]));
        var regexp = Get(new String(Value.AsString(arguments[1])));
        var match = regexp.Match(s);
        if (match.Success)
        {
            var groups = new object[match.Groups.Count];
            for (var i = 0; i < match.Groups.Count; i++)
            {
                groups[i] = match.Groups[i].Value.ToCharArray();
            }
            return Pair.List(groups);
        }
        return Value.F;
    }
}
